src/derivers/bip39.ts

```diff
@@ -1,17 +1,91 @@
-import { hmac } from '@noble/hashes/hmac';
-import { pbkdf2 } from '@noble/hashes/pbkdf2';
-import { sha512 } from '@noble/hashes/sha512';
+import { assert, stringToBytes } from '@metamask/utils';
 
 import type { DeriveChildKeyArgs } from '.';
 import type { BIP39StringNode } from '../constants';
 import { BYTES_KEY_LENGTH } from '../constants';
+import type { CryptographicFunctions } from '../cryptography';
+import { hmacSha512, pbkdf2Sha512 } from '../cryptography';
 import type { Curve } from '../curves';
 import { SLIP10Node } from '../SLIP10Node';
 import { getFingerprint } from '../utils';
 
+const MNEMONIC_PHRASE_LENGTHS = [12, 15, 18, 21, 24];
+
+/**
+ * Validate a BIP-39 mnemonic phrase. The phrase must:
+ *
+ * - Consist of 12, 15, 18, 21, or 24 words.
+ * - Contain only words from the English wordlist.
+ *
+ * @param mnemonicPhrase - The mnemonic phrase to validate.
+ * @throws If the mnemonic phrase is invalid.
+ */
+function validateMnemonicPhrase(mnemonicPhrase: string) {
+  const words = mnemonicPhrase.split(' ');
+
+  assert(
+    MNEMONIC_PHRASE_LENGTHS.includes(words.length),
+    `Invalid mnemonic phrase: The mnemonic phrase must consist of 12, 15, 18, 21, or 24 words.`,
+  );
+
+  assert(
+    words.every((word) => englishWordlist.includes(word)),
+    'Invalid mnemonic phrase: The mnemonic phrase contains an unknown word.',
+  );
+}
+
+/**
+ * Encode a BIP-39 mnemonic phrase to a `Uint8Array` for use in seed generation.
+ * If the mnemonic is already a `Uint8Array`, it is assumed to contain the
+ * indices of the words in the wordlist.
+ *
+ * @param mnemonic - The mnemonic phrase to encode.
+ * @param wordlist - The wordlist to use.
+ * @returns The encoded mnemonic phrase.
+ */
+function encodeMnemonicPhrase(
+  mnemonic: string | Uint8Array,
+  wordlist: string[],
+) {
+  if (typeof mnemonic === 'string') {
+    validateMnemonicPhrase(mnemonic);
+    return stringToBytes(mnemonic.normalize('NFKD'));
+  }
+
+  const mnemonicString = Array.from(new Uint16Array(mnemonic.buffer))
+    .map((i) => wordlist[i])
+    .join(' ');
+
+  validateMnemonicPhrase(mnemonicString);
+  return stringToBytes(mnemonicString);
+}
+
+/**
+ * Convert a BIP-39 mnemonic phrase to a seed.
+ *
+ * @param mnemonic - The BIP-39 mnemonic phrase to convert. If the mnemonic is a
+ * `Uint8Array`, it is assumed to contain the indices of the words in the
+ * English wordlist.
+ * @param passphrase - The passphrase to use.
+ * @param cryptographicFunctions - The cryptographic functions to use. If
+ * provided, these will be used instead of the built-in implementations.
+ */
+export async function mnemonicToSeed(
+  mnemonic: string | Uint8Array,
+  passphrase = '',
+  cryptographicFunctions?: CryptographicFunctions,
+) {
+  return await pbkdf2Sha512(
+    encodeMnemonicPhrase(mnemonic, englishWordlist),
+    stringToBytes(`mnemonic${passphrase}`.normalize('NFKD')),
+    2048,
+    64,
+    cryptographicFunctions,
+  );
+}
```

Comment:
```suggestion
  const salt = `mnemonic${passphrase}`.normalize('NFKD');
  return await pbkdf2Sha512(
    encodeMnemonicPhrase(mnemonic, englishWordlist),
    stringToBytes(salt),
    2048,
    64,
    cryptographicFunctions,
  );
}
```
Potentially worth it for readability